075-auto-rebalancing-cluster-scaling.md

```diff
@@ -0,0 +1,435 @@
+3. On the **Nth** reconciliation (rebalancing id done, scaling down can start):
+     * the check on scaling down the cluster doesn't fail, because the auto-rebalancing was completed and the brokers to be removed don't host any topics' partitions anymore.
+     * the `scaleDown()` method can run the scaling down.
+     * the `KafkaAutorebalancingReconciler().reconcile()` detects that the `KafkaRebalance` custom resource status reports the `Ready` state (so the rebalancing is completed). The end of rebalancing is reflected into the `Kafka.status.autoRebalance.state` field and the `KafkaRebalance` is deleted.
+
+Because the rebalancing runs across several reconciliations, different scenarios could happen even taking into account operator restarts:
+
+* the `removedNodes` list is empty:
+  * it means no scale down requested. Nothing to do in terms of rebalancing.
+* the `removedNodes` list is not empty:
+  * the `Kafka.status.autoRebalance.state` doesn't exist or it's `Ready`. It's the normal flow, a new rebalance has to start.
+  * the `Kafka.status.autoRebalance.state` is at `RebalancingOnScaleDown`. The removed brokers' IDs are compared with the `Kafka.status.autoRebalance.modes.brokers` field. If they are equal, it's the normal flow, so just checking the rebalancing state. If they are different, the user could have asked for another scale down right after a previous one. The operator has to stop the current rebalancing, by applying the `strimzi.io/rebalance: stop` annotation on the corresponding `KafkaRebalance` custom resource, and start a new rebalancing, by applying the `strimzi.io/rebalance: refresh` annotation, with a consistent list of brokers defined by the multiple requested scales down.
+
+During the above flow, errors can happen before the scaling down:
+
+* There is an error when getting the proposal.
+* After auto-approving the proposal, there is an error when running the rebalance.
+
+In both scenarios, the reconciliation ends with the `KafkaRebalance` custom resource in the `NotReady` state which is reflected into the `Kafka.status.autoRebalance.state` field.
+
+The reconcile ends but the scale down cannot proceed because the brokers to be removed are still hosting some partition replicas.
+On the next reconciliation, the cluster operator will try to do a scale down again, so going through the same flow.
+The scale down check fails but the rebalancing cannot run because of the `KafkaRebalance` custom resource in the `NotReady` state.
+The user can fix it by deleting the `KafkaRebalance` custom resource so that on the next reconciliation, the operator will be able to retry an auto-rebalancing, by recreating a new `KafkaRebalance` custom resource.
+Another approach the user can take is removing the `spec.cruiseControl.autoRebalance` (other than deleting the `KafkaRebalance` custom resource) in order to run a manual rebalancing later on.
```

Comment:
What happens if the rebalance failed, but the scaledown is not needed anymore (was reverted by the user or the rebalance was done out-of-band)? Will the operator detect that this NotReady rebalance is not needed anymore and reset itself?